### PR TITLE
Add RuleKind to RuleConfiguration

### LINKED ISSUE #1282: Rules should explain their reason for existing
For example, why does the `empty_count` rule exist? Is it purely a stylistic preference? Is the compiler able to better optimize `isEmpty`?

Why do we have `shorthand_operator`? Is there some official Swift guide somewhere that states shortand syntax is more idiomatic?

These are the kinds of things I'd like to know when choosing what rules to disable or opt-in. This could potentially lead to less disabled rules in general.

The simplest thing would be for rules to be able to include a rationale field that can be viewed somehow, likely using the `rules` command.

If we wanted to get a little more fancy, maybe rules could be annotated with attributes such as `stylistic` for rules that are purely subjective, and provide no objective benefit either way. `idiomatic` could be used for rules that are considered idiomatic Swift. `stability` might be applied to rules that avoid stability issues such as `weak_delegate`.

### PR DESCRIPTION
**Pull request by marcelofabri**
Partially fixes #1282

In some of the rules I was in doubt between 2 kinds, so this can/should be validated.

### CODE REVIEW

**Comment by jpsim** [comment]
I definitely like this direction, that's the right way to do this IMO./

**Comment by norio-nomura** [comment]
> In some of the rules I was in doubt between 2 kinds, so this can/should be validated.

If Kind would be protocol, rule could be adapt to multiple kinds.

**Comment by marcelofabri** [comment]
> If Kind would be protocol, rule could adapt to multiple kinds.

My initial feeling is that from an user perspective it's better that a rule can belong to only one kind, even if some of them are in a "gray area".

File: Source/SwiftLintFramework/Protocols/Rule.swift [hunk 1]
```diff
@@ -8,10 +8,19 @@
...
+    case style
+    case metrics
+    case performance
+}
+
 public protocol Rule {
     init() // Rules need to be able to be initialized with default values
     init(configuration: Any) throws
     static var description: RuleDescription { get }
+    static var kind: RuleKind { get }
```
Comment (masters3d) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: What about using an option set? I think the advantage is that a rule can be classifed with multiple "labels" without those labels having to be protocols. I like the idea of protocols as labels too. 
The emum restriction kind of reminds me of  http://www.jessesquires.com/enums-as-configs/

Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: I still think that from an user point of view, it's better that a rule can only belong to one category.

This way we could easily in the future allow enabling/disabling full categories that once. That would be tricky if a rule was in two groups: what would be the expected behavior?

**Comment by marcelofabri** [comment]
@jpsim Is this something that we want to move forward with? 

Revisiting this, I'd add `kind` in `RuleConfiguration` instead.

File: Source/SwiftLintFramework/Protocols/Rule.swift (hunk 1 above)
Comment (masters3d) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: What about config options that change behavior?

Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: I'm not sure I understand, can you elaborate?

Comment (masters3d) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: Rules with configurations that make the rule be a different category depending on what config is given.

Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: Can you think of any examples?

Comment (masters3d) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: Colon, Comma.

Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/Rule.swift]: for me they'd be style rules no matter the configuration

**Comment by marcelofabri** [comment]
I've rebased this and moved the definition to `RuleDescription`. Basically now we should:

- [ ] Validate that the kinds (do they meaningfully divide the rules?)
- [ ] Validate each rule kind
- [ ] Choose what to do with custom rules. Currently it's always flagged as `.style`. Options:
    * Allow `kind` to be specified in configuration? (What if it's not specified?)
    * Create a new kind `.custom`?

**Pull request merged**